PurchasesCoreSwift/SubscriberAttributes/SubscriberAttributesManager.swift

```diff
@@ -0,0 +1,263 @@
+            let receivedNSError = error as NSError?
+            Logger.error(String(format: Strings.attribution.attributes_sync_error,
+                                receivedNSError?.localizedDescription ?? "",
+                                receivedNSError?.userInfo ?? ""))
+        }
+    }
+
+    @objc public func unsyncedAttributesByKey(appUserID: String) -> SubscriberAttributeDict {
+        // TODO  these logs were in RCPurchases+subscriberattributes, are we ok with logging
+        // them every time?
+        let unsyncedAttributes = deviceCache.unsyncedAttributesByKey(appUserID: appUserID)
+        Logger.debug(String(format: Strings.attribution.unsynced_attributes_count,
+                            unsyncedAttributes.count,
+                            appUserID))
+        if unsyncedAttributes.count > 0 {
+            Logger.debug(String(format: Strings.attribution.unsynced_attributes, unsyncedAttributes))
+        }
+
+        return unsyncedAttributes
+    }
+
+    @objc public func unsyncedAttributesByKeyForAllUsers() -> [String: SubscriberAttributeDict] {
+        return deviceCache.unsyncedAttributesForAllUsers()
+    }
+
+    @objc public func markAttributesAsSynced(_ maybeAttributesToSync: SubscriberAttributeDict?, appUserID: String) {
+        guard let attributesToSync = maybeAttributesToSync,
+              !attributesToSync.isEmpty else {
+            return
+        }
+
+        Logger.info(String(format: Strings.attribution.marking_attributes_synced,
+                           appUserID,
+                           attributesToSync.description))
+
+        lock.lock()
+        var unsyncedAttributes = unsyncedAttributesByKey(appUserID: appUserID)
+        for (key, attribute) in attributesToSync {
+            if let unsyncedAttribute = unsyncedAttributes[key] {
+                if unsyncedAttribute == attribute {
+                    unsyncedAttribute.isSynced = true
+                    unsyncedAttributes[key] = unsyncedAttribute
+                }
+            }
+        }
+        deviceCache.store(subscriberAttributesByKey: unsyncedAttributes, appUserID: appUserID)
+        lock.unlock()
+    }
+
+    @objc public func convertAttributionDataAndSetAsSubscriberAttributes(attributionData: [String: Any],
+                                                                         network: AttributionNetwork,
+                                                                         appUserID: String) {
+        let convertedAttribution = attributionDataMigrator.convertToSubscriberAttributes(
+            attributionData: attributionData,
+            network: network.rawValue)
+        setAttributes(convertedAttribution, appUserID: appUserID)
+    }
+
+    private func storeAttributeLocallyIfNeeded(key: String, value: String?, appUserID: String) {
+        let currentValue = currentValueForAttribute(key: key, appUserID: appUserID)
+        if currentValue == nil || currentValue != value.rc_orEmpty() {
```

Comment:
FFTI: I think a guard let return would be slightly more idiomatic

Reply:
since we are continuing if it IS null, i thought that'd get a little complicated

Reply:
i thought that'd get a little complicated since we continue if current value IS nil... a quick google didn't show a cool way to do that, but maybe i'm just not searching for the right thing?